### PR TITLE
Arena followup

### PR DESCRIPTION
**Pull request by quetz**
- huge allocs (separate list, stuff that does not fit our blocks go there)
- some more helper funcs
- slices (ptr+len+cap) types to be passed around by value
- unit tests

- [x] Mark this if you consider it ready to merge
- [x] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

### CODE REVIEW

File: libr/include/r_util/r_arena.h [hunk 1]
```diff
@@ -9,12 +9,27 @@
...
 
+// Helper type to pass ptr+len around by value
+typedef struct r_slice_t {
+  const uint8_t *ptr;
+  size_t len;
+  size_t cap;
+} RSlice;
+
+R_API RSlice r_slice(const void *ptr, size_t len);
+R_API RSlice r_empty_slice(); // or just RSlice x = { 0 };
```
Comment (trufae) [inline libr/include/r_util/r_arena.h]:
```suggestion
R_API RSlice r_empty_slice(void); // or just RSlice x = { 0 };
```

File: libr/util/arena.c [hunk 2]
```diff
@@ -9,6 +9,46 @@ _Static_assert (sizeof (size_t) >= sizeof (void *), "size_t must fit pointer");
...
+
+R_API size_t r_slice_len(RSlice slice) {
+  return slice.len;
+}
+
+R_API const uint8_t *r_slice_ptr(RSlice slice) {
+  return slice.ptr;
+}
+
+R_API RSlice r_empty_slice() {
```
Comment (trufae) [inline libr/util/arena.c]:
```suggestion
R_API RSlice r_empty_slice(void) {
```

File: libr/util/arena.c [hunk 3]
```diff
@@ -9,6 +9,46 @@ _Static_assert (sizeof (size_t) >= sizeof (void *), "size_t must fit pointer");
...
+  return (RSlice) { .ptr = slice.ptr, .len = to, .cap = slice.cap };
+}
+
+R_API RSlice r_slice_from(RSlice slice, size_t from) {
+  if (from >= slice.len) return r_empty_slice();
+  return (RSlice) { .ptr = slice.ptr + from, .len = slice.len - from, .cap = slice.cap };
+}
+
+R_API RSlice r_slice_from_to(RSlice slice, size_t from, size_t to) {
+  if (from >= slice.len || from > to) return r_empty_slice();
```
Comment (trufae) [inline libr/util/arena.c]:
```suggestion
  if (from >= slice.len || from > to) {
   return r_empty_slice();
  }
```

File: libr/util/arena.c [hunk 4]
```diff
@@ -9,6 +9,46 @@ _Static_assert (sizeof (size_t) >= sizeof (void *), "size_t must fit pointer");
...
+  if (to >= slice.len) to = slice.len;
```
Comment (trufae) [inline libr/util/arena.c]: braces here

File: libr/util/arena.c [hunk 5]
```diff
@@ -58,6 +102,7 @@ R_API RArena *r_arena_create_with(size_t block_size) {
 	arena->block_size = block_size;
 	arena->total_allocated = 0;
 	arena->default_alignment = ARENA_DEFAULT_ALIGNMENT;
+  arena->huge = NULL;
```
Comment (trufae) [inline libr/util/arena.c]: bad indentation.. just run clang-format-radare2 in libr/util/arena.c

File: test/unit/test_arena.c [hunk 6]
```diff
@@ -0,0 +1,583 @@
...
+
+	// Edge case: to = 0
+	result = r_slice_to(slice, 0);
+	mu_assert_eq(result.len, 0, "slice_to(0) should create empty slice");
+	mu_assert_ptreq(result.ptr, data, "slice_to(0) should preserve ptr");
+
+	mu_end;
+}
+
+bool test_r_slice_from(void) {
```
Comment (trufae) [inline test/unit/test_arena.c]: make all those tests functions static inline, and also run clang-format-radare2 in them

**Review by trufae — CHANGES_REQUESTED** [review]
aa

File: libr/util/arena.c [hunk 7]
```diff
@@ -259,3 +384,66 @@ R_API void *r_arena_push(RArena *arena, const void *mem, size_t n) {
 	}
 	return copy;
 }
+
+// Copy arbitrary memory into arena and free it
+R_API void *r_arena_move(RArena *arena, void *mem, size_t n) {
+	if (!arena || !mem || n == 0) {
```
Comment (trufae) [inline libr/util/arena.c]: Those should be R_RETURN_VAL_IF_FAIL

File: test/unit/test_arena.c [hunk 8]
```diff
@@ -0,0 +1,567 @@
...
+	mu_assert_notnull (slice.ptr, "slice should be allocated");
+	mu_assert_eq (slice.len, 100, "slice length should be 100");
+
+	// Test scalloc (slice calloc)
+	RSlice zslice = r_arena_scalloc (arena, 50);
+	mu_assert_notnull (zslice.ptr, "zeroed slice should be allocated");
+	mu_assert_eq (zslice.len, 50, "zeroed slice length should be 50");
+
+	bool all_zero = true;
+	for (size_t i = 0; i < zslice.len; i++) {
```
Comment (trufae) [inline test/unit/test_arena.c]:
```suggestion
	size_t i;
	for (i = 0; i < zslice.len; i++) {
```

File: libr/include/r_util/r_arena.h [hunk 9]
```diff
@@ -31,27 +46,42 @@ typedef struct r_arena_block_t {
 typedef struct r_arena_t {
 	RArenaBlock *current; // Current block we're allocating from
 	RArenaBlock *first; // First block (kept for reset)
+
```
Comment (trufae) [inline libr/include/r_util/r_arena.h]:
```suggestion

```

File: libr/include/r_util/r_arena.h [hunk 10]
```diff
@@ -31,27 +46,42 @@ typedef struct r_arena_block_t {
 typedef struct r_arena_t {
 	RArenaBlock *current; // Current block we're allocating from
 	RArenaBlock *first; // First block (kept for reset)
+                      
+  RArenaBlock *huge; // Requests larger than block size go here (1 block = 1 alloc)
```
[the remark this answers is not in the document]
Comment (trufae) [reply libr/include/r_util/r_arena.h]: indent with tabs pls

File: libr/util/arena.c [hunk 11]
```diff
@@ -9,6 +9,51 @@ _Static_assert (sizeof (size_t) >= sizeof (void *), "size_t must fit pointer");
 _Static_assert (sizeof (uintptr_t) >= sizeof (void *),
 	"uintptr_t must fit pointer");
 
+R_API RSlice r_slice(const void *ptr, size_t len) {
```
Comment (trufae) [inline libr/util/arena.c:12]: as said, all those functions must be defined as static inline in the .h

File: libr/util/arena.c [hunk 12]
```diff
@@ -9,6 +9,51 @@ _Static_assert (sizeof (size_t) >= sizeof (void *), "size_t must fit pointer");
...
+	}
+	if (to >= slice.len) {
+		to = slice.len;
+	}
+
+	return (RSlice){ .ptr = slice.ptr + from, .len = to - from, .cap = slice.cap };
+}
+
 static inline uintptr_t align_address(uintptr_t address, size_t alignment) {
 	assert (alignment > 0 && (alignment &(alignment - 1)) == 0);
```
Comment (trufae) [inline libr/util/arena.c:58]: do not use assert. use the R_RETURN macros instead

File: libr/include/r_util/r_arena.h [hunk 13]
```diff
@@ -9,12 +9,27 @@
 extern "C" {
 #endif
 
+// Helper type to pass ptr+len around by value
+typedef struct r_slice_t {
+  const uint8_t *ptr;
```
Comment (trufae) [inline libr/include/r_util/r_arena.h:14]: use tabs not spaces
